File: fabric-groups/src/main/scala/org/fusesource/fabric/groups/Group.scala
/**
 * Copyright (C) 2010-2011, FuseSource Corp.  All rights reserved.
 *
 *     http://fusesource.com
 *
 * The software in this package is published under the terms of the
 * CDDL license a copy of which has been included with this distribution
 * in the license.txt file.
 */
package org.fusesource.fabric.groups

import scala.reflect.BeanProperty
import org.linkedin.zookeeper.client.IZKClient
import org.fusesource.fusemq.cluster.ZooKeeperGroup
/**
 * <p>
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
object GroupFactory {
  def create(zk: IZKClient, path: String):Group = new ZooKeeperGroup(zk, path)
  def members(zk: IZKClient, path: String):Array[Member] = ZooKeeperGroup.members(zk, path)
}

/**
 * <p>
 *   Used the join a cluster group and to monitor the memberships
 *   of that group.
 * </p>
 * <p>
 *   This object is not thread safe.  You should are responsible for
 *   synchronizing access to it across threads.
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
trait Group {

  /**
   * Adds a member to the group with some associated data.  Both the
   * id and the data will be visible to anyone listing the members of the
   * group.
   *
   * If you join
   */
  def join(id:String, data:Array[Byte]):Unit

  /**
   * Removes a previously added member.
   */
  def leave(id:String):Unit

  /**
   * Lists all the members currently in the group.
   */
  def members():Array[Member]

  /**
   * Registers a change listener which will be called
   * when the cluster membership changes.
   */
  def add(listener:ChangeListener)

  /**
   * Removes a previously added change listener.
   */
  def remove(listener:ChangeListener)

  /**
   * A group should be closed to release aquired resources used
   * to monitor the group membership.
   *
   * Whe the Group is closed, any memberships registered via this
   * Group will be removed from the group.
   */
  def close:Unit

}

/**
 * <p>
 *   Callback interface used to get notifications of changes
 *   to a cluster group.
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
trait ChangeListener {

  /**
   * Processing this change even is not allowed to block for very long.
   */
  def changed(members:Array[Member]):Unit
}

/**
 * <p>
 *   Represents a member in the cluster group.
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
case class Member(
   @BeanProperty
   id:String,
   @BeanProperty
   data:Array[Byte]
)

File: fabric/fabric-groups/src/main/scala/org/fusesource/fabric/groups2/Group.java
/**
 * Copyright (C) FuseSource, Inc.
 * http://fusesource.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.fusesource.fabric.groups2;

import java.io.Closeable;
import java.io.IOException;
import java.util.List;
import java.util.Map;

public interface Group<T extends NodeState> extends Closeable {

    /**
     * Are we connected with the cluster?
     */
    boolean isConnected();

    /**
     * Start this member
     */
    void start();

    /**
     * A member should be closed to release acquired resources used
     * to monitor the group membership.
     *
     * When the member is closed, any memberships registered via this
     * Group will be removed from the group.
     */
    void close() throws IOException;

    /**
     * Registers a change listener which will be called
     * when the cluster membership changes.
     */
    void add(GroupListener<T> listener);

    /**
     * Removes a previously added change listener.
     */
    void remove(GroupListener<T> listener);

    /**
     * Update the state of this group member.
     * If the state is null, the member will leave the group.
     *
     * @param state the new state of this group member
     */
    void update(T state);

    Map<String, T> members();

    boolean isMaster();

    T master();

    List<T> slaves();

}
